tensorboard/webapp/runs/store/utils.ts

```diff
@@ -41,7 +41,58 @@ export function groupRuns(
       break;
 
     case GroupByKey.REGEX:
-      throw new Error('Not implemented');
+      if (!groupBy.regexString) {
+        //TODO(japie1235813): props users when the input is invalid
+        throw new Error('Empty regex string.');
+      }
+
+      // TODO(japie1235813): add additonal `\` to convert string to regex, which
+      // makes `new RegExp()` construct properly
+      // For example, convert `foo\d+bar` to `foo\\d+bar`
+
+      try {
+        new RegExp(groupBy.regexString);
+      } catch (e) {
+        //TODO(japie1235813): props users when the input is invalid
+        throw new Error('Invalid regex.');
+      }
+
+      const regExp = new RegExp(groupBy.regexString);
+      // Checks if there is capture group in regex.
+      let isCaptureGroup = false;
+
+      for (const run of runs) {
+        let matches = run.name.match(regExp);
+        if (matches) {
+          if (matches.length > 1) {
+            matches = matches.slice(1);
+            isCaptureGroup = true;
+          }
+          const id = matches.length === 1 ? matches[0] : matches.join('_');
```

Comment:
thanks for pointing out `matches.join` case
although I don't see why it's a bad matches when the matches includes `'_'`, I agree that `JSON.stringify` is a better option. updated.

Reply:
In case the regex string is something like `foo(_|)bar(_|)`, the matches for `foo_bar` and `foobar_` would be both:
- `foo_bar`: `["foo_bar", "_", ""]`
- `foobar_`: `["foo_bar", "", "_"]`

and,
`["_", ""].join("_") === "__"` and `["", "_"].join("_") === "__"` which would be faulty.

Reply:
I see. thanks for the explanation.